storage/postgres/query_builder.go

```diff
@@ -0,0 +1,330 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/jmoiron/sqlx"
+)
+
+type orderRule struct {
+	field     string
+	orderType query.OrderType
+}
+
+type QueryBuilder struct {
+	db pgDB
+
+	sql         string
+	queryParams []interface{}
+
+	labelCriteria, fieldCriteria []query.Criterion
+	orderByFields                []orderRule
+	limit                        int
+	criteria                     []query.Criterion
+	hasLock                      bool
+	returningFields              []string
+
+	err error
+}
+
+func NewQueryBuilder(db pgDB) *QueryBuilder {
+	return &QueryBuilder{
+		db: db,
+	}
+}
+
+func (qb *QueryBuilder) List(ctx context.Context, entity PostgresEntity) (*sqlx.Rows, error) {
+	if qb.err != nil {
+		return nil, qb.err
+	}
+
+	baseTableName := entity.TableName()
+	var baseQuery string
+	label := entity.LabelEntity()
+	if label == nil {
+		baseQuery = constructBaseQueryForEntity(baseTableName)
+	} else {
+		baseQuery = constructBaseQueryForLabelable(label, baseTableName)
+	}
+	qb.sql = baseQuery
+
+	if err := qb.finalizeSQL(entity); err != nil {
+		return nil, err
+	}
+
+	return qb.db.QueryxContext(ctx, qb.sql, qb.queryParams...)
+}
+
+func (qb *QueryBuilder) Delete(ctx context.Context, entity PostgresEntity) (*sqlx.Rows, error) {
+	if qb.err != nil {
+		return nil, qb.err
+	}
+	baseTableName := entity.TableName()
+	qb.sql = fmt.Sprintf("DELETE FROM %s", baseTableName)
+	for len(qb.labelCriteria) > 0 {
+		return nil, &util.UnsupportedQueryError{Message: "conditional delete is only supported for field queries"}
+	}
+
+	if err := qb.finalizeSQL(entity); err != nil {
+		return nil, err
+	}
+	return qb.db.QueryxContext(ctx, qb.sql, qb.queryParams...)
+}
+
+func (qb *QueryBuilder) Return(fields ...string) *QueryBuilder {
+	qb.returningFields = append(qb.returningFields, fields...)
+
+	return qb
+}
+
+func (qb *QueryBuilder) Limit(limit int) *QueryBuilder {
```

Comment:
These 2 should be private. You should configure these only via `WithCriteria`

Reply:
It is nice to have these, as they allow for easier query building

Reply:
Removed them